Propagating parametricBrother to node descendants

IpaliboWhyte opened the pull request:
Ref: https://github.com/delvedor/find-my-way/issues/104

- Updated previousPath to use outstanding characters that have been trimmed off
- Propagating **parametricBrother** to node descendants
- Added tests

mcollina commented on node.js:
```diff
@@ -72,14 +72,24 @@ Node.prototype.addChild = function (node) {
-    if (child.kind === this.types.STATIC && parametricBrother) {
-      child.parametricBrother = parametricBrother
+  // Save the parametric brother inside static children
+  let iterate = (node) => {
```
this should be a `const`.

IpaliboWhyte replied:
Done :)

delvedor commented on index.js:
```diff
@@ -385,6 +384,10 @@ Router.prototype.find = function find (method, path, version) {
       }
+
+      const tempPath = originalPath.slice(0, originalPath.length - path.length)
+      const previousPath = `${tempPath.slice(tempPath.lastIndexOf('/') + 1, tempPath.length)}${path}`
```
Can you use `var` instead of `const`?
```suggestion
      var previousPath = `${tempPath.slice(tempPath.lastIndexOf('/') + 1, tempPath.length)}${path}`
```
Same in the line above.

delvedor commented on test/issue-104.test.js:
```diff
@@ -0,0 +1,57 @@
+const test = t.test
+const FindMyWay = require('../')
+
+test('Parametric route, url with parameter common prefix > 1', t => {
```
Can you also add a test with the same code of the issue? :)

IpaliboWhyte replied:
Added, test called `Nested static parametric route, url with parameter common prefix > 1` :)

IpaliboWhyte commented:
So I discovered another [issue](https://github.com/delvedor/find-my-way/issues/104#issuecomment-447418032) which I found to be also related to **parametricBrother**. Previously when fetching the first parametric route, it ignored the fact that the node may already have a parametricBrother defined. So I added a fix to reassign.

I also added some more complex scenarios in tests, I'm also open to more suggestions :D

**Benchmarks**

**issue-104:**

```
lookup static route x 22,498,921 ops/sec ±1.91% (77 runs sampled)
lookup dynamic route x 2,614,556 ops/sec ±3.10% (85 runs sampled)
lookup dynamic multi-parametric route x 1,506,951 ops/sec ±1.52% (83 runs sampled)
lookup dynamic multi-parametric route with regex x 1,144,684 ops/sec ±0.89% (86 runs sampled)
lookup long static route x 2,651,074 ops/sec ±1.16% (87 runs sampled)
lookup long dynamic route x 2,060,354 ops/sec ±1.29% (81 runs sampled)
lookup static versioned route x 5,337,143 ops/sec ±1.50% (86 runs sampled)
find static route x 30,166,017 ops/sec ±3.46% (78 runs sampled)
find dynamic route x 3,361,503 ops/sec ±1.01% (86 runs sampled)
find dynamic multi-parametric route x 1,854,761 ops/sec ±1.08% (87 runs sampled)
find dynamic multi-parametric route with regex x 1,351,089 ops/sec ±1.00% (85 runs sampled)
find long static route x 3,266,728 ops/sec ±1.69% (84 runs sampled)
find long dynamic route x 2,521,191 ops/sec ±1.18% (83 runs sampled)
find static versioned route x 6,401,119 ops/sec ±1.15% (81 runs sampled)
```

**Master**

```
lookup static route x 23,253,994 ops/sec ±1.27% (84 runs sampled)
lookup dynamic route x 2,752,674 ops/sec ±1.34% (82 runs sampled)
lookup dynamic multi-parametric route x 1,537,593 ops/sec ±1.27% (86 runs sampled)
lookup dynamic multi-parametric route with regex x 1,137,470 ops/sec ±1.45% (82 runs sampled)
lookup long static route x 2,770,461 ops/sec ±1.01% (89 runs sampled)
lookup long dynamic route x 2,129,942 ops/sec ±1.45% (82 runs sampled)
lookup static versioned route x 5,887,928 ops/sec ±1.22% (88 runs sampled)
find static route x 32,495,521 ops/sec ±1.25% (83 runs sampled)
find dynamic route x 3,402,508 ops/sec ±1.22% (85 runs sampled)
find dynamic multi-parametric route x 1,848,192 ops/sec ±1.35% (80 runs sampled)
find dynamic multi-parametric route with regex x 1,298,883 ops/sec ±1.13% (89 runs sampled)
find long static route x 3,233,918 ops/sec ±1.38% (85 runs sampled)
find long dynamic route x 2,418,767 ops/sec ±1.84% (87 runs sampled)
find static versioned route x 6,379,556 ops/sec ±1.55% (84 runs sampled)
```

IpaliboWhyte commented on index.js:
```diff
@@ -385,6 +384,13 @@ Router.prototype.find = function find (method, path, version) {
+      // e.g originalPath: /aa/bbb/cc, path: bb/cc
+      // outstanding path: /bbb/cc
+      const pathDiff = originalPath.slice(0, originalPath.length - path.length)
+      const previousPath = `${pathDiff.slice(pathDiff.lastIndexOf('/') + 1, pathDiff.length)}${path}`
```
Ahh yes, fixed.

IpaliboWhyte commented on test/issue-104.test.js:
```diff
@@ -0,0 +1,145 @@
+const t = require('tap')
```
Done

IpaliboWhyte commented on index.js:
```diff
@@ -385,6 +384,13 @@ Router.prototype.find = function find (method, path, version) {
+      // we need to know the outstanding path so far from the originalPath since the last encountered "/" and assign it to previousPath.
+      // e.g originalPath: /aa/bbb/cc, path: bb/cc
+      // outstanding path: /bbb/cc
+      const pathDiff = originalPath.slice(0, originalPath.length - path.length)
```
`originalPath.length` now cached in `originalPathLength`.
Also why should we use `var` here? :)

delvedor replied:
Because var is faster than const due to the scoping :)
(And for being consistent with the current code)

mcollina commented on index.js:
```diff
@@ -385,6 +385,13 @@ Router.prototype.find = function find (method, path, version) {
+      // e.g originalPath: /aa/bbb/cc, path: bb/cc
+      // outstanding path: /bbb/cc
+      const pathDiff = originalPath.slice(0, originalPathLength - pathLen)
+      const previousPath = pathDiff.slice(pathDiff.lastIndexOf('/') + 1, pathDiff.length) + path
```
Can you check if a fast-path for this possible? I would prefer if we could detect the condition that lead to the issue being fixed by this PR, and then apply these checks instead?

IpaliboWhyte replied:
Good point, just added a check to verify the `previousPath` is "complete" since the last "/". 
This seems to be the condition as we keep trimming `path` which could lead to `previousPath` becoming one or more characters short.

mcollina commented:
Does the new if clause improve the benchmarks?

IpaliboWhyte commented:
Benchmarks

**issue-104:**
```
lookup static route x 46,363,939 ops/sec ±1.55% (90 runs sampled)
lookup dynamic route x 3,282,410 ops/sec ±1.55% (90 runs sampled)
lookup dynamic multi-parametric route x 1,786,782 ops/sec ±1.24% (89 runs sampled)
lookup dynamic multi-parametric route with regex x 1,405,629 ops/sec ±0.38% (89 runs sampled)
lookup long static route x 3,206,773 ops/sec ±0.50% (91 runs sampled)
lookup long dynamic route x 2,249,853 ops/sec ±0.88% (92 runs sampled)
lookup static versioned route x 8,584,778 ops/sec ±0.46% (90 runs sampled)
find static route x 34,585,327 ops/sec ±0.71% (90 runs sampled)
find dynamic route x 4,054,133 ops/sec ±0.50% (92 runs sampled)
find dynamic multi-parametric route x 2,156,954 ops/sec ±0.26% (93 runs sampled)
find dynamic multi-parametric route with regex x 1,529,215 ops/sec ±1.69% (93 runs sampled)
find long static route x 4,731,089 ops/sec ±0.74% (91 runs sampled)
find long dynamic route x 3,080,737 ops/sec ±1.02% (93 runs sampled)
find static versioned route x 11,071,305 ops/sec ±0.55% (92 runs sampled)
```

**Master**
```
lookup static route x 47,357,726 ops/sec ±0.72% (93 runs sampled)
lookup dynamic route x 3,298,239 ops/sec ±1.26% (93 runs sampled)
lookup dynamic multi-parametric route x 1,689,911 ops/sec ±2.05% (89 runs sampled)
lookup dynamic multi-parametric route with regex x 1,324,307 ops/sec ±1.68% (91 runs sampled)
lookup long static route x 3,083,328 ops/sec ±1.38% (89 runs sampled)
lookup long dynamic route x 2,205,086 ops/sec ±0.90% (88 runs sampled)
lookup static versioned route x 8,623,795 ops/sec ±0.40% (91 runs sampled)
find static route x 35,487,898 ops/sec ±0.34% (95 runs sampled)
find dynamic route x 4,222,883 ops/sec ±0.77% (89 runs sampled)
find dynamic multi-parametric route x 2,169,592 ops/sec ±0.73% (91 runs sampled)
find dynamic multi-parametric route with regex x 1,491,537 ops/sec ±1.47% (93 runs sampled)
find long static route x 4,768,656 ops/sec ±0.45% (92 runs sampled)
find long dynamic route x 3,120,219 ops/sec ±1.11% (94 runs sampled)
find static versioned route x 11,245,763 ops/sec ±0.52% (92 runs sampled)
```

mcollina commented on node.js:
```diff
@@ -82,7 +82,9 @@ Node.prototype.addChild = function (node) {
-    node.parametricBrother = parametricBrother
+    if (node !== this) {
+      node.parametricBrother = parametricBrother
+    }
```
Can you add a unit test for this change? When would this happen

IpaliboWhyte replied:
This would happen when `iterate(this)` [is invoked](https://github.com/delvedor/find-my-way/pull/105/files#diff-28a3689be95c88dd5e7a37db516ab084R95). `this` remains constant meaning `node === this` on the first call. We want to ignore that case and just recursively iterate it's children.

mcollina replied:
+1, can you add a unit test, or is it already included?

IpaliboWhyte replied:
I will try add a test case for that soon :)

IpaliboWhyte replied:
Test case Added :)
